### PR TITLE
Restructure augmentation package

### LINKED ISSUE #1514: Restructuring of the augmentation package
## 🚀 Feature

This would introduce a cleaner interface making it much easier to work on a single thing.

## Motivation

I struggled a lot when I first went into the source code of kornia, especially the augmentation and as someone coming from other languages than python it is still weird to see multiple classes per file.
Moreover I think files with more than thousand lines might indicate bad style.

## Pitch

Reorder files into `2d` and `3d` subfolders which allows to easily spot, whats supported.
Outsource each class into a single file wherever it is possible and makes sense.
To not break too much, the external interface of augmentation should not change.

## Alternatives

Keep as is.

## Additional context

cc @shijianjian 
______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

### PR DESCRIPTION
**Pull request by twsl**
#### Changes

Fixes #1514

Outsources augmentations into respective `2d` or `3d` folders with separated files.
I sneaked in a devcontainer file to simplify my development.

#### Type of change

- [ ] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Did you update CHANGELOG in case of a major change?

cc @edgarriba @shijianjian

### CODE REVIEW

**Comment by shijianjian** [comment]
Any better ideas handling those base classes? espectially for the mix augmentations. I feel like to somehow differ them from other 2D augmentations

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# nvidia/cuda:11.4.1-cudnn8-devel-ubuntu20.04
```
Comment (edgarriba) [inline .devcontainer/Dockerfile:1]: file not needed

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,97 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
```
Comment (edgarriba) [inline .devcontainer/devcontainer.json:1]: file not needed

File: kornia/augmentation/_2d/geometric/affine.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
+from typing import Dict, Optional, Tuple, Union, cast
+
+import torch
+
+from kornia.augmentation import random_generator as rg
+from kornia.augmentation._2d.geometric.base import GeometricAugmentationBase2D
```
Comment (edgarriba) [inline kornia/augmentation/_2d/geometric/affine.py:6]: I think that would be interesting to expose `GeometricAugmentationBase2D` to enable kornia users to easiliy create their custom stuff 

```python
from kornia.augmentation import GeometricAugmentationBase2D
```

File: kornia/augmentation/_2d/__init__.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+from kornia.augmentation._2d.geometric import *
```
Comment (edgarriba) [inline kornia/augmentation/_2d/__init__.py:1]: why `_2d` and not `2d` ?

File: kornia/augmentation/_2d/geometric/horizontal_flip.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        )
+
+        return flip_mat.repeat(input.size(0), 1, 1)
+
+    def apply_transform(
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor], transform: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        return hflip(input)
+
+    def inverse_transform(
```
Comment (edgarriba) [inline kornia/augmentation/_2d/geometric/horizontal_flip.py:70]: @shijianjian why here we overload the function and not actually inverting the transform?

File: pyproject.toml [hunk 6]
```diff
@@ -2,3 +2,18 @@
 line-length = 120
 skip-string-normalization = true
 skip-magic-trailing-comma = true
+
+[tool.isort]
+multi_line_output=3
+include_trailing_comma = true
+line_length = 120
+profile = "black"
+known_first_party = [
```
Comment (edgarriba) [inline pyproject.toml:11]: what's doing ?

**Review by edgarriba — COMMENTED** [review]
added few comments and questions. There are some issues with testing possible could be solved here @shijianjian

File: .devcontainer/Dockerfile (hunk 1 above)
Comment (twsl) [reply .devcontainer/Dockerfile:1]: Ideally I'd like to use https://github.com/microsoft/vscode-dev-containers/pull/1176 cause the current devcontainer only supports CUDA on windows through WSL2

File: kornia/augmentation/_2d/geometric/affine.py (hunk 3 above)
Comment (twsl) [reply kornia/augmentation/_2d/geometric/affine.py:6]: Yes, I already exposed that

File: pyproject.toml (hunk 6 above)
Comment (twsl) [reply pyproject.toml:11]: added the isort config as isort recommends pyproject for configuration. But didn't remove it from setup.cfg as i didnt want to break any existing workflow

File: kornia/augmentation/_2d/__init__.py (hunk 4 above)
Comment (twsl) [reply kornia/augmentation/_2d/__init__.py:1]: throws invalid syntax error. IIRC, python package names must start with a literal or underscore

Comment (edgarriba) [reply kornia/augmentation/_2d/__init__.py:1]: just out of curiosity, could you share link to that ?

Comment (twsl) [reply kornia/augmentation/_2d/__init__.py:1]: [good ol' stackoverflow](https://stackoverflow.com/a/17487228) with valid links to [identifiers](https://docs.python.org/3/reference/lexical_analysis.html#identifiers) and [grammar-token-module](https://docs.python.org/3/reference/simple_stmts.html#grammar-token-module) for Python 3

File: kornia/augmentation/_2d/geometric/horizontal_flip.py (hunk 5 above)
Comment (shijianjian) [reply kornia/augmentation/_2d/geometric/horizontal_flip.py:70]: The inversion happened internally. The `inverse_transform` is about how to use the inverse matrix.

**Comment by edgarriba** [comment]
@twsl @shijianjian any idea why the tests are failing ?

**Comment by twsl** [comment]
@edgarriba i spent some time investigating, but i couldnt find the cause yet.
I think the bounding boxes do not respecting the padding.
It was probably introduced by the combination of104a5d783788bd5e37338d721e4caf506b0508 and 3f8e7a34384b3a6671861b2777682df0c28cc8c4 already

File: kornia/augmentation/container/augment.py [hunk 7]
```diff
@@ -1,23 +1,25 @@
 import warnings
 from itertools import zip_longest
-from typing import Any, cast, List, Optional, Tuple, Union
+from typing import Any, List, Optional, Tuple, Union, cast
 
 import torch
 
-from kornia.augmentation.base import _AugmentationBase, GeometricAugmentationBase2D, IntensityAugmentationBase2D
+from kornia.augmentation._2d.geometric.base import GeometricAugmentationBase2D
+from kornia.augmentation._2d.intensity.base import IntensityAugmentationBase2D
```
Comment (shijianjian) [inline kornia/augmentation/container/augment.py:8]: since it started with the underscore, so I would suggest to hide these packages
```suggestion
from kornia.augmentation import GeometricAugmentationBase2D
from kornia.augmentation import IntensityAugmentationBase2D
```

File: kornia/augmentation/container/base.py [hunk 8]
```diff
@@ -4,7 +4,8 @@
 import torch
 import torch.nn as nn
 
-from kornia.augmentation.base import _AugmentationBase, MixAugmentationBase
+from kornia.augmentation._2d.mix.base import MixAugmentationBase
```
Comment (shijianjian) [inline kornia/augmentation/container/base.py:7]:
```suggestion
from kornia.augmentation import MixAugmentationBase
```

File: kornia/augmentation/container/image.py [hunk 9]
```diff
@@ -6,16 +6,12 @@
...
-    IntensityAugmentationBase2D,
-    MixAugmentationBase,
-    TensorWithTransformMat,
-)
-
-from .base import ParamItem, SequentialBase
-from .utils import ApplyInverseInterface, InputApplyInverse
+from kornia.augmentation._2d.geometric.base import GeometricAugmentationBase2D
+from kornia.augmentation._2d.intensity.base import IntensityAugmentationBase2D
+from kornia.augmentation._2d.mix.base import MixAugmentationBase
```
Comment (shijianjian) [inline kornia/augmentation/container/image.py:11]:
```suggestion
from kornia.augmentation import GeometricAugmentationBase2D
from kornia.augmentation import IntensityAugmentationBase2D
from kornia.augmentation import MixAugmentationBase
```

File: kornia/augmentation/container/patch.py [hunk 10]
```diff
@@ -4,12 +4,12 @@
 import torch
 import torch.nn as nn
 
-from kornia.augmentation.base import _AugmentationBase, MixAugmentationBase, TensorWithTransformMat
+from kornia.augmentation._2d.mix.base import MixAugmentationBase
```
Comment (shijianjian) [inline kornia/augmentation/container/patch.py:7]:
```suggestion
from kornia.augmentation import MixAugmentationBase
```

**Review by shijianjian — APPROVED** [review]

**Review by edgarriba — APPROVED** [review]
